File: 19. Remove Nth Node From End of List/19. Remove Nth Node From End of List.cpp
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        ListNode* first = dummy;
        ListNode* second = dummy;
        
        // Move first pointer n nodes ahead
        for (int i = 0; i <= n; i++) {
            first = first->next;
        }
        
        // Move both pointers until first reaches the end
        while (first != nullptr) {
            first = first->next;
            second = second->next;
        }
        
        // Remove the nth node
        ListNode* temp = second->next;
        second->next = second->next->next;
        delete temp;
        
        return dummy->next;
    }
};

File: 19. Remove Nth Node From End of List/19. Remove Nth Node From End of List.go
func removeNthFromEnd(head *ListNode, n int) *ListNode {
    dummy := &ListNode{0, head}
    first := dummy
    second := dummy
    
    // Move first pointer n nodes ahead
    for i := 0; i <= n; i++ {
        first = first.Next
    }
    
    // Move both pointers until first reaches the end
    for first != nil {
        first = first.Next
        second = second.Next
    }
    
    // Remove the nth node
    second.Next = second.Next.Next
    
    return dummy.Next
}